commands/audit/sca/gem/gem.go

```diff
@@ -0,0 +1,460 @@
+	}
+
+	gemGraphInfoContent, err := parseGemfileLockDeps(lockFilePath)
+	if err != nil {
+		err = fmt.Errorf("error generating gem graph JSON from Gemfile.lock: %w", err)
+		return
+	}
+	var gemInput GemGraphInput
+	if err = json.Unmarshal(gemGraphInfoContent, &gemInput); err != nil {
+		err = fmt.Errorf("failed to unmarshal gem graph JSON: %w", err)
+		return
+	}
+	if len(gemInput.Graph.Nodes) == 0 {
+		log.Debug("No gem dependencies found in GemGraphInput after parsing Gemfile.lock.")
+		return []*xrayUtils.GraphNode{}, []string{}, nil
+	}
+
+	projectRootNode := buildFullGemDependencyGraph(gemInput)
+
+	if projectRootNode != nil && projectRootNode.Nodes != nil {
+		dependencyTrees = projectRootNode.Nodes
+	} else {
+		dependencyTrees = []*xrayUtils.GraphNode{}
+	}
+	uniqueDeps = calculateUniqueDependencies(dependencyTrees)
+
+	log.Debug("Calculated dependency trees (children of root): %d trees found.", len(dependencyTrees))
+
+	return
+}
+
+// parseLockfileToInternalData parses a Gemfile.lock file line by line to extract gem specifications and their dependencies.
+func parseLockfileToInternalData(lockFilePath string) (
+	orderedGems []*internalGemRef,
+	resolvedVersions map[string]string,
+	err error,
+) {
+	file, ioErr := os.Open(lockFilePath)
+	if ioErr != nil {
+		return nil, nil, fmt.Errorf("opening lockfile %s: %w", lockFilePath, ioErr)
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+
+	if !advanceToSpecs(scanner) {
+		return []*internalGemRef{}, make(map[string]string), nil
+	}
+
+	orderedGems, resolvedVersions = parseSpecsSection(scanner)
+
+	if scanErr := scanner.Err(); scanErr != nil {
+		return nil, nil, fmt.Errorf("error scanning lockfile: %w", scanErr)
+	}
+
+	return orderedGems, resolvedVersions, nil
+}
+
+// advanceToSpecs moves the scanner to the line immediately following the "specs:" heading.
+// It returns true if the section is found, and false otherwise.
+func advanceToSpecs(scanner *bufio.Scanner) bool {
+	foundGemBlock := false
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		if !foundGemBlock {
+			if line == "GEM" {
+				foundGemBlock = true
+			}
+			continue
+		}
+		if line == "specs:" {
+			return true
+		}
+	}
+	return false // Reached end of file without finding the section.
+}
+
+// parseSpecsSection processes the lines within the "specs:" block of the lockfile.
+// It uses indentation levels to distinguish between gems and their dependencies.
+func parseSpecsSection(scanner *bufio.Scanner) (
+	orderedGems []*internalGemRef,
+	resolvedVersions map[string]string,
+) {
+	orderedGems = []*internalGemRef{}
+	resolvedVersions = make(map[string]string)
+	var currentGem *internalGemRef
+
+	for scanner.Scan() {
+		line := scanner.Text()
+		trimmedLine := strings.TrimSpace(line)
+		if trimmedLine == "" {
+			currentGem = nil
+			continue
+		}
+
+		if sectionTerminators[trimmedLine] {
+			break
+		}
+
+		indentation := countLeadingSpaces(line)
+
+		switch indentation {
+		case 4:
```

Comment:
comment on each case, this code is not clear why `4`, `6` are the needed cases and what they represent